europi.go

```diff
@@ -5,60 +5,107 @@ import (
+	B2Pin = machine.GPIO5
+
+	CV1Pin = machine.GPIO21
+	CV2Pin = machine.GPIO20
+	CV3Pin = machine.GPIO16
+	CV4Pin = machine.GPIO17
+	CV5Pin = machine.GPIO18
+	CV6Pin = machine.GPIO19
+)
+
+var (
+	DisplayChannel = machine.I2C0
+	CV1PwmGroup    = machine.PWM2
+	CV2PwmGroup    = machine.PWM2
+	CV3PwmGroup    = machine.PWM0
+	CV4PwmGroup    = machine.PWM0
+	CV5PwmGroup    = machine.PWM1
+	CV6PwmGroup    = machine.PWM1
+
+	europi *EuroPi
 )
 
 // EuroPi is the collection of component wrappers used to interact with the module.
 type EuroPi struct {
-	// Display is a wrapper around ssd1306.Device
-	Display *Display
+	// Display provides methods for drawing to the OLED display.
+	Display *display
+
+	DI *digitalInput
+	// AI provides methods for reading analog input control voltage between 0 and 12V.
+	AI *analogInput
 
-	DI DigitalReader
-	AI AnalogReader
+	// B1 is a struct for handling the left push button behavior.
+	B1 *button
+	// B2 is a struct for handling the right push button behavior.
+	B2 *button
 
-	B1 DigitalReader
-	B2 DigitalReader
+	// K1 provides methods for reading knob voltage and position for the left knob.
+	K1 *knob
+	// K2 provides methods for reading knob voltage and position for the left knob.
+	K2 *knob
 
-	K1 AnalogReader
-	K2 AnalogReader
+	// CV1-6 are structs for interacting with the cv output jacks.
+	CV1 *output
+	CV2 *output
+	CV3 *output
+	CV4 *output
+	CV5 *output
+	CV6 *output
+	// CV is an array containing all CV outputs.
+	CV [6]*output
+}
 
-	CV1 Outputer
-	CV2 Outputer
-	CV3 Outputer
-	CV4 Outputer
-	CV5 Outputer
-	CV6 Outputer
-	CV  [6]Outputer
+func init() {
+	europi = new()
 }
 
-// New will return a new EuroPi struct.
-func New() *EuroPi {
-	cv1 := NewOutput(machine.GPIO21, machine.PWM2)
-	cv2 := NewOutput(machine.GPIO20, machine.PWM2)
-	cv3 := NewOutput(machine.GPIO16, machine.PWM0)
-	cv4 := NewOutput(machine.GPIO17, machine.PWM0)
-	cv5 := NewOutput(machine.GPIO18, machine.PWM1)
-	cv6 := NewOutput(machine.GPIO19, machine.PWM1)
+func new() *EuroPi {
+	cv1 := newOutput(CV1Pin, CV1PwmGroup)
+	cv2 := newOutput(CV2Pin, CV2PwmGroup)
+	cv3 := newOutput(CV3Pin, CV3PwmGroup)
+	cv4 := newOutput(CV4Pin, CV4PwmGroup)
+	cv5 := newOutput(CV5Pin, CV5PwmGroup)
+	cv6 := newOutput(CV6Pin, CV6PwmGroup)
 
 	return &EuroPi{
-		Display: NewDisplay(machine.I2C0, machine.GPIO0, machine.GPIO1),
+		Display: newDisplay(DisplayChannel, DisplaySdaPin, DisplaySclPin),
 
-		DI: NewDI(machine.GPIO22),
-		AI: NewAI(machine.ADC0),
+		DI: newDigitalInput(DIPin),
+		AI: newAnalogInput(AIPin),
 
-		B1: NewButton(machine.GPIO4),
-		B2: NewButton(machine.GPIO5),
+		B1: newButton(B1Pin),
+		B2: newButton(B2Pin),
 
-		K1: NewKnob(machine.ADC1),
-		K2: NewKnob(machine.ADC2),
+		K1: newKnob(K1Pin),
+		K2: newKnob(K2Pin),
 
 		CV1: cv1,
 		CV2: cv2,
 		CV3: cv3,
 		CV4: cv4,
 		CV5: cv5,
 		CV6: cv5,
-		CV:  [6]Outputer{cv1, cv2, cv3, cv4, cv5, cv6},
+		CV:  [6]*output{cv1, cv2, cv3, cv4, cv5, cv6},
 	}
 }
+
+func GetInstance() *EuroPi {
```

Comment:
OK, so when I said singletons are great... Ummm... is there a thing as taking things too far?

Hahaha, kidding aside this pattern is questionable from a configuration and mock-test standpoint. Since you only have **one** EuroPi ever and more importantly no ways to initialize it there's no way to mock `EuroPi` and also no way to configure it either. Consider returning to the `New()` function and letting the user say when they want to start the `EuroPi`.

The reason I draw the line on singletons on the EuroPi is because the EuroPi type is not a direct abstraction on hardware as is a PWM, I2C and such, but rather a type that holds the direct HW abstractions and operates on them. There is no reason for `EuroPi` to be a singleton really unless you really want it to be a singleton, which is OK, I guess.